tiny_dnn/core/kernels/fully_connected_op_avx.h

```diff
@@ -14,225 +14,302 @@ namespace kernels {
 
 #ifdef CNN_USE_AVX
 
-template <typename Allocator>
-inline void avx_fully_connected_forward_kernel(
-  const std::vector<std::vector<float, Allocator>> &in_data,
-  const std::vector<float, Allocator> &W,
-  const std::vector<float, Allocator> &bias,
-  std::vector<std::vector<float, Allocator>> &out_data,
-  const fully_params &params,
-  const bool layer_parallelize) {
-  if (params.has_bias_) {
-    size_t nblocks  = params.out_size_ / 8;
-    size_t nremains = params.out_size_ & 7;
+/**
+ * Kernel for forward propogation for fully connected layer with AVX backend
+ * (single precision)
+ * @param in_data
+ * @param W
+ * @param bias
+ * @param out_data
+ * @param params
+ * @param layer_parallelize
+ */
+template <typename S1, typename S2, typename S3, typename S4>
+inline void avx_fully_connected_forward_kernel(const Tensor<float, S1> &in_data,
+                                               const Tensor<float, S2> &W,
+                                               const Tensor<float, S3> &bias,
+                                               Tensor<float, S4> &out_data,
+                                               const bool has_bias,
+                                               const bool layer_parallelize) {
+  size_t out_size = out_data.shape()[1], in_size = in_data.shape()[1];
+  if (has_bias) {
+    size_t nblocks  = out_size / 8;
+    size_t nremains = out_size & 7;
     if (nremains) {
       int32_t mask_src[] = {
         -1, -1, -1, -1, -1, -1, -1, -1, 0, 0, 0, 0, 0, 0, 0, 0,
       };
       __m256i imask =
         _mm256_loadu_si256((__m256i const *)(mask_src + 8 - nremains));
-      for_i(layer_parallelize, in_data.size(), [&](int sample) {
-        const auto &in = in_data[sample];
-        auto &out      = out_data[sample];
+      for_i(layer_parallelize, in_data.shape()[0], [&](int sample) {
+        auto in  = in_data.host_iter(sample, 0);
+        auto out = out_data.host_iter(sample, 0);
```

Comment:
These local variables should be pointers instead of iterators.
You may use `Tensor::host_pointer` method. (Though I wish we can skip creating iterator instance by not using `host_iter` method in there...)